src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1453,42 +1453,102 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
                     }
                 }
 
-                DiagnosticBag warnings = DiagnosticBag.GetInstance();
-                var viableEntryPoints = ArrayBuilder<MethodSymbol>.GetInstance();
+                var intOrVoidEntryPoints = ArrayBuilder<MethodSymbol>.GetInstance();
+                // Validity and diagnostics are also tracked because they must be conditionally handled 
+                // if there are not any "traditional" entrypoints found.
+                var taskEntryPoints = ArrayBuilder<(bool IsValid, MethodSymbol Candidate, DiagnosticBag SpecificDiagnostics)>.GetInstance();
+
+                // These diagnostics (warning only) are added to the compilation only if 
+                // there were not any main methods found.
+                DiagnosticBag noMainFoundDiagnostics = DiagnosticBag.GetInstance();
+
                 foreach (var candidate in entryPointCandidates)
                 {
-                    if (!HasEntryPointSignature(candidate, warnings))
+                    var perCandidateBag = DiagnosticBag.GetInstance();
+                    var (IsCandidate, IsTaskLike) = HasEntryPointSignature(candidate, perCandidateBag);
+
+                    if (IsCandidate && !IsTaskLike)
                     {
-                        // a single error for partial methods
-                        warnings.Add(ErrorCode.WRN_InvalidMainSig, candidate.Locations.First(), candidate);
-                        continue;
+                        intOrVoidEntryPoints.Add(candidate);
+                        perCandidateBag.Free();
+                    }
+                    else if (IsTaskLike)
+                    {
+                        taskEntryPoints.Add((IsCandidate, candidate, perCandidateBag));
+                    }
+                    else
+                    {
+                        noMainFoundDiagnostics.Add(ErrorCode.WRN_InvalidMainSig, candidate.Locations.First(), candidate);
+                        noMainFoundDiagnostics.AddRange(perCandidateBag);
+                        perCandidateBag.Free();
                     }
```

Comment:
Perhaps check `IsTaskLike` first. That would simplify the `IsCandidate` check and `perCandidateBag` could be freed in one place rather than two.
```
if (IsTaskLike)
{
    taskEntryPoints.Add(...);
}
else
{
    if (IsCandidate)
    {
        intOrVoidEntryPoints.Add(candidate);
    }
    else
    {
        noMainFoundDiagnostics.Add(...);
    }
    perCandidatesBag.Free();
}